TELCODOCS-504 - Validating cluster tuning for vDU application workloads (#45949)
openshift/openshift-docs: modules/ztp-du-cluster-config-reference.adoc

:_module-type: REFERENCE
[id="ztp-du-cluster-config-reference_{context}"]
= Recommended cluster configurations to run vDU applications

Clusters running virtualized distributed unit (vDU) applications require a highly tuned and optimized configuration. The following information describes the various elements that you require to support vDU workloads in {product-title} {product-version} clusters.

== Recommended cluster MachineConfig CRs

The following `MachineConfig` CRs configure the cluster host:

.Recommended MachineConfig CRs
[cols=2*, width="80%", options="header"]
|====
|CR filename
|Description

|`02-workload-partitioning.yaml`
|Configures workload partitioning for the cluster. Apply this `MachineConfig` CR when you install the cluster.

|`MachineConfigSctp.yaml`
|Loads the SCTP kernel module. This `MachineConfig` CR is optional and can be omitted if you do not require this kernel module.

|`MachineConfigContainerMountNS.yaml`
|Configures the container mount namespace and kubelet conf.

|`MachineConfigAcceleratedStartup.yaml`
|Configures accelerated startup for the cluster.

|`06-kdump-master.yaml`, `06-kdump-worker.yaml`

> |Configures kdump for the cluster.

Reviewer: should kdump be in back ticks? I've seen it both ways. In this context it might not be.

|====

== Recommended cluster Operators

The following Operators are required for clusters running vDU applications and are a part of the baseline reference configuration:

> . Node Tuning Operator (NTO). NTO packages functionality that was previously delivered with the Performance Addon Operator, which is now a part of NTO.

Reviewer: Should we number or bullet point here? Maybe it doesn't matter.

. PTP Operator
. SR-IOV Network Operator
. Red Hat OpenShift Logging Operator
. Local Storage Operator

== Recommended cluster kernel configuration

Always use the latest supported realtime kernel version in your cluster. You should also ensure that the following configurations are applied in the cluster:

> . Ensure the following `additionalKernelArgs` are set in the cluster performance profile:

Reviewer: "By running the following command."

Author: @stevsmit ;) this isn't a command...

// Module included in the following assemblies:
//
// * scalability_and_performance/ztp-vdu-configuration-reference.adoc

:_module-type: REFERENCE
[id="ztp-du-cluster-config-reference_{context}"]
= Recommended cluster configurations to run vDU applications

> Clusters running distributed unit (DU) applications require a highly tuned and optimized configuration. The following information describes the various elements that you require to deploy a vDU in {product-title} {product-version} clusters.

Reviewer 2: ```suggestion
Clusters running distributed unit (DU) applications require a highly tuned and optimized configuration. The following information describes the various elements that you require to support vDU workloads in {product-title} {product-version} clusters.
```

[NOTE]
====
You manage your custom site configuration data in a Git repository accessible from the hub cluster that is defined as a source repository for Argo CD. You extract the baseline source custom resources (CRs) from the zero touch provisioning (ZTP) container.
====

== Recommended cluster MachineConfig installation CRs

The following `MachineConfig` install CRs configure the cluster host:

.MachineConfig installation CRs
[cols=2*, width="80%", options="header"]
|====
|CR filename
|Description

> |`03-workload-partitioning.yaml`

Reviewer 2: ```suggestion
|`02-workload-partitioning.yaml`
```

|Configures workload partitioning for the cluster.

|`MachineConfigSctp.yaml`

> |Configures SCTP in the cluster.

Reviewer 2: ```suggestion
|Loads the SCTP kernel module. This MachineConfig is optional and may be omitted if the use case does not require this kernel module.
```

|`MachineConfigContainerMountNS.yaml`
|Configures the container mount namespace and kubelet conf.

|`MachineConfigAcceleratedStartup.yaml`
|Configures accelerated startup for the cluster.

|`06-kdump-master.yaml`, `06-kdump-worker.yaml`
|Configures kdump for the cluster.
|====

== Recommended cluster Operators

The following Operators are required for clusters running vDU applications and are installed when the cluster is provisioned with the ZTP pipeline:

. Node Tuning Operator (NTO). NTO packages functionality that was previously delivered with the Performance Addon Operator, which is now a part of NTO.
. PTP Operator
. SR-IOV Network Operator
. Red Hat OpenShift Logging Operator
. Local Storage Operator

== Recommended cluster kernel configuration

Always use the latest supported realtime kernel version in your cluster. You should also ensure that the following configurations are applied in the cluster:

. Ensure the following `additionalKernelArgs` are set in the cluster performance profile:
+
[source,yaml]
----
spec:
  additionalKernelArgs:
  - "idle=poll"

>   - "rcupdate.rcu_normal_after_boot=0"

Reviewer 2: ```suggestion
  - "rcupdate.rcu_normal_after_boot=0"
  - "efi=runtime"
```
The `efi=runtime` argument is necessary if SecureBoot is enabled.

// Module included in the following assemblies:
//
// * scalability_and_performance/ztp-vdu-configuration-reference.adoc

:_module-type: REFERENCE
[id="ztp-du-cluster-config-reference_{context}"]
= Recommended cluster configurations to run vDU applications

Clusters running virtualized distributed unit (vDU) applications require a highly tuned and optimized configuration. The following information describes the various elements that you require to support vDU workloads in {product-title} {product-version} clusters.

[NOTE]
====
You manage your custom site configuration data in a Git repository accessible from the hub cluster that is defined as a source repository for Argo CD. You extract the baseline source custom resources (CRs) from the zero touch provisioning (ZTP) container.

> ====

Reviewer 2: Minor note... I think we can leave this note out. This doc is focused on validating that the correct configuration is applied. How that configuration is applied is up to the customer (even if we strongly recommend using a gitops approach).

Author: Removed.

> == Recommended cluster MachineConfig installation CRs

Reviewer 2: Minor note... I would leave out the "installation" qualifier. It implies these are somehow different than typical MachineConfig CRs. (3 total places in this section)

The following `MachineConfig` installation CRs configure the cluster host:

.MachineConfig installation CRs
[cols=2*, width="80%", options="header"]
|====
|CR filename
|Description

|`02-workload-partitioning.yaml`

> |Configures workload partitioning for the cluster.

Reviewer 2: By leaving out "installation" in the title, we should make a note that this MachineConfig must be applied at installation time.

|`MachineConfigSctp.yaml`
|Loads the SCTP kernel module. This `MachineConfig` CR is optional and can be omitted if you do not require this kernel module.

|`MachineConfigContainerMountNS.yaml`
|Configures the container mount namespace and kubelet conf.

|`MachineConfigAcceleratedStartup.yaml`
|Configures accelerated startup for the cluster.

|`06-kdump-master.yaml`, `06-kdump-worker.yaml`
|Configures kdump for the cluster.
|====

== Recommended cluster Operators

> The following Operators are required for clusters running vDU applications and are installed when the cluster is provisioned with the ZTP pipeline:

Reviewer 2: Minor edit (removing the "how" things are applied):
```suggestion
The following Operators are required for clusters running vDU applications and are part of the baseline reference configuration:
```

. PTP Operator
. SR-IOV Network Operator
. Red Hat OpenShift Logging Operator
. Local Storage Operator

== Recommended cluster kernel configuration

Always use the latest supported realtime kernel version in your cluster. You should also ensure that the following configurations are applied in the cluster:

. Ensure the following `additionalKernelArgs` are set in the cluster performance profile:
+
[source,yaml]
----
spec:
  additionalKernelArgs:
  - "idle=poll"
  - "rcupdate.rcu_normal_after_boot=0"
  - "efi=runtime"
----

. Ensure that the `performance-patch` profile in the `Tuned` CR configures the correct CPU isolation set that matches the `isolated` CPU set in the related `PerformanceProfile` CR, for example:
+
[source,yaml]
----
spec:
  profile:
    - name: performance-patch
      # The 'include' line must match the associated PerformanceProfile name
      # And the cmdline_crash CPU set must match the 'isolated' set in the associated PerformanceProfile
      data: |
        [main]
        summary=Configuration changes profile inherited from performance created tuned
        include=openshift-node-performance-openshift-node-performance-profile
        [bootloader]

>         cmdline_crash=nohz_full=2-51,54-103 <1>

Reviewer 3: This does not match with the config in performance profile. 
Either change performanceprofile configs to 0-1,52-53, or change this config to 3-39,43-79. 
@imiller0 do we prefer to configure 6 or 4 reserved cpus?

Reviewer 2: Prefer 4 to match up with wording elsewhere. 4 is the minimum necessary reserved with additional possibly needed based on the users workload.

Author: I have updated examples based on the following CPU allocations:

```
  cpu:
    isolated: 0-1,52-53
    reserved: 2-4,40-42
```
